File: test-suite-kotlin/src/main/kotlin/example/micronaut/service/MovieServiceImpl.kt
package example.micronaut.service

import example.micronaut.configuration.AppConfiguration
import example.micronaut.exception.MovieServiceException
import jakarta.inject.Singleton
import org.opensearch.client.opensearch.OpenSearchClient
import org.opensearch.client.opensearch._types.FieldValue
import org.opensearch.client.opensearch._types.query_dsl.MatchQuery
import org.opensearch.client.opensearch._types.query_dsl.Query
import org.opensearch.client.opensearch.core.IndexRequest
import org.opensearch.client.opensearch.core.IndexResponse
import org.opensearch.client.opensearch.core.SearchRequest
import org.opensearch.client.opensearch.core.SearchResponse
import org.opensearch.client.opensearch.core.search.Hit
import org.slf4j.Logger
import org.slf4j.LoggerFactory

@Singleton
class MovieServiceImpl(
    private val appConfiguration: AppConfiguration,
    private val client: OpenSearchClient
) : MovieService {

    override fun saveMovie(movie: Movie): String {
        try {
            val indexRequest: IndexRequest<Movie> = createIndexRequest(movie)
            val indexResponse: IndexResponse = client.index(indexRequest)
            val id = indexResponse.id()
            LOG.info("Document for '{}' {} successfully in ES. The id is: {}", movie, indexResponse.result(), id)
            return id
        } catch (e: Exception) {
            val errorMessage = String.format("An exception occurred while indexing '%s'", movie)
            LOG.error(errorMessage)
            throw MovieServiceException(errorMessage, e)
        }
    }

    private fun createIndexRequest(movie: Movie) = IndexRequest.Builder<Movie>()
            .index(appConfiguration.moviesIndexName)
            .document(movie)
            .build()

    override fun searchMovies(title: String): Movie? {
        try {
            val searchResponse: SearchResponse<Movie> =
                client.search({ s: SearchRequest.Builder ->
                    s.index(appConfiguration.moviesIndexName)
                        .query { q: Query.Builder ->
                            q.match { m: MatchQuery.Builder ->
                                m.field("title")
                                    .query { fq: FieldValue.Builder ->
                                        fq.stringValue(
                                            title
                                        )
                                    }
                            }
                        }
                }, Movie::class.java)
            LOG.info(
                "Searching for '{}' took {} and found {}",
                title,
                searchResponse.took(),
                searchResponse.hits().total().value()
            )
            val hits: Iterator<Hit<Movie>> = searchResponse.hits().hits().iterator()
            if (hits.hasNext()) {
                return hits.next().source()
            }
            return null
        } catch (e: Exception) {
            val errorMessage = String.format("An exception occurred while searching for title '%s'", title)
            LOG.error(errorMessage)
            throw MovieServiceException(errorMessage, e)
        }
    }

    companion object {
        private val LOG: Logger = LoggerFactory.getLogger(MovieServiceImpl::class.java)
    }
}

File: test-suite-python/src/test/python/micronaut/example/service/MovieServiceImpl.py
import logging

from jakarta.inject import Singleton
from java.lang import Exception as JavaException
from org.opensearch.client.opensearch import OpenSearchClient
from org.opensearch.client.opensearch.core import IndexRequest

from micronaut.example.configuration.AppConfiguration import AppConfiguration
from micronaut.example.exception.MovieServiceException import MovieServiceException
from micronaut.example.service.Movie import Movie
from micronaut.example.service.MovieService import MovieService

LOG = logging.getLogger(__name__)


@Singleton
class MovieServiceImpl(MovieService):

    def __init__(self, app_configuration: AppConfiguration, client: OpenSearchClient):
        self.app_configuration = app_configuration
        self.client = client

    def save_movie(self, movie: Movie) -> str:
        try:
            index_request = self.create_index_request(movie)
            index_response = self.client.index(index_request)
            id = index_response.id()
            LOG.info("Document for '%s' %s successfully in ES. The id is: %s", movie, index_response.result(), id)
            return id
        except JavaException as e:
            error_message = f"An exception occurred while indexing '{movie}'"
            LOG.error(error_message)
            raise MovieServiceException(error_message, e)

    def create_index_request(self, movie: Movie) -> IndexRequest[Movie]:
        return (IndexRequest.Builder()
                .index(self.app_configuration.movies_index_name)
                .document(movie)
                .build())

    def search_movies(self, title: str) -> Movie | None:
        try:
            search_response = self.client.search(
                lambda s: s.index(self.app_configuration.movies_index_name).query(
                    lambda q: q.match(
                        lambda m: m.field("title").query(lambda fq: fq.stringValue(title)))),
                Movie)
            LOG.info("Searching for '%s' took %s and found %s",
                     title,
                     search_response.took(),
                     search_response.hits().total().value())
            hits = search_response.hits().hits()
            if not hits.isEmpty():
                return hits.get(0).source()
            return None
        except JavaException as e:
            error_message = f"An exception occurred while searching for title '{title}'"
            LOG.error(error_message)
            raise MovieServiceException(error_message, e)
